### PR TITLE
Add new Base32 implementation

### PR DESCRIPTION
**Pull request by adolfogc**
## Description
Provides a port of the new Base32 implementation, plus fixes to make it compatible with Google Authenticator's implementation, that was previously introduced in PRs #984 and #1001.

## Motivation and context
See issue #1022.

## How has this been tested?
Manually on Linux.

## Screenshots (if appropriate):
N/A.

## Types of changes

- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ I have added tests to cover my changes.

### CODE REVIEW

**Comment by adolfogc** [comment]
I've rebased.

File: src/core/Base32.cpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+        while (n >= 0) {
+            int index = (quantum & mask) >> n;
+            encodedData[o++] = alphabet[index];
+            mask >>= 5;
+            n -= 5;
+        }
+
+        // add pad characters
+        while (o < encodedData.size())
```
Comment (TheZ3ro) [inline src/core/Base32.cpp]: add brackets

File: src/core/Optional.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#define OPTIONAL_H
+
+/*
+ * This utility class is for providing basic support for an option type.
+ * It can be replaced by std::optional (C++17) or
+ * std::experimental::optional (C++11) when they become fully supported
+ * by all the main compiler toolchains.
+ */
+
+template <typename T> class Optional
```
Comment (phoerious) [inline src/core/Optional.h]: Do we really need our own Optional implementation? Would the same be possible with QVariant?

File: src/totp/totp.cpp [hunk 3]
```diff
@@ -93,30 +91,53 @@ QString QTotp::parseOtpString(QString key, quint8 &digits, quint8 &step)
...
     code.addData(QByteArray(reinterpret_cast<char*>(&current), sizeof(current)));
     QByteArray hmac = code.result();
 
     int offset = (hmac[hmac.length() - 1] & 0xf);
-    int binary =
-            ((hmac[offset] & 0x7f) << 24)
-            | ((hmac[offset + 1] & 0xff) << 16)
-            | ((hmac[offset + 2] & 0xff) << 8)
-            | (hmac[offset + 3] & 0xff);
+    int binary = ((hmac[offset] & 0x7f) << 24) | ((hmac[offset + 1] & 0xff) << 16) | ((hmac[offset + 2] & 0xff) << 8) |
```
Comment (phoerious) [inline src/totp/totp.cpp]: I liked the previous formatting better. It's hard to read on a single line.

File: src/core/Base32.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    QByteArray data(nBytes, Qt::Uninitialized);
+
+    int i = 0;
+    int o = 0;
+
+    while (i < encodedData.size()) {
+        quint64 quantum = 0;
+        int nQuantumBytes = 5;
+
+        for (int n = 0; n < 8; n++) {
```
[the remark this answers is not in the document]
Comment (adolfogc) [reply src/core/Base32.cpp]: Ok, will change it.

File: src/core/Optional.h (hunk 2 above)
Comment (adolfogc) [reply src/core/Optional.h]: I think we can make this change to reduce the code base size, as the variant type can serve as an optional.

File: src/totp/totp.cpp (hunk 3 above)
Comment (adolfogc) [reply src/totp/totp.cpp]: I agree with you, this was the result of running clang-format; I'll change it.

File: src/core/Optional.h (hunk 2 above)
Comment (phoerious) [reply src/core/Optional.h]: Just just QVariant::isNull() or QVariant::isValid().

Comment (adolfogc) [reply src/core/Optional.h]: I'm almost done making this change, I decided for `QVariant::isNull`; as I understand that`QVariant::isValid` is used for detecting when a type not defined for the variant (i.e. not registered with Q_DECLARE_METATYPE) is used.

File: src/core/Base32.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+            quantum <<= 5;
+            quantum |= ch;
+        }
+
+        const int offset = (nQuantumBytes - 1) * 8;
+        quint64 mask = quint64(0xFF) << offset;
+        for (int n = offset; n >= 0; n -= 8) {
+            char c = static_cast<char>((quantum & mask) >> n);
+            data[o++] = c;
```
[the remark this answers is not in the document]
Comment (adolfogc) [reply src/core/Base32.cpp]: Good catch! This helped me find a bug in `Base32::decode()` affecting how the number of bytes of decoded data was being calculated, even though it didn't truncate the result, it was causing the array to be resized unnecessarily. I will be submitting the changes soon.

**Comment by adolfogc** [comment]
@phoerious I've finished making the changes solicited, please review them.

File: src/core/Base32.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (phoerious) [reply src/core/Base32.cpp]: Yay!

**Comment by phoerious** [comment]
Nice work. Thanks!

**Pull request merged**